internal/shim-sgx/src/enclave.rs

```diff
@@ -38,280 +125,63 @@ mod _internal {
-    mov     rax,                    4
-    enclu
-
-    jmp     {RET_FROM_SYSCALL}
-    ",
-        RET_FROM_SYSCALL = sym ret_from_syscall,
-        options(noreturn)
-        )
-    }
-
     #[no_mangle]
-    #[naked]
-    pub unsafe extern "sysv64" fn ret_from_syscall() -> ! {
-        asm!("
-# rax = syscall return stack pointer
-# rbx = next non-enclave instruction
-# rcx = &TCS
-# r10 = untrusted stack pointer
-# r11 = &aex[CSSA - 1]
-# rsp = trusted stack pointer
-    mov     QWORD PTR [r11 + {SRSP}], 0             # Clear syscall return stack pointer field
-    mov     rsp,                    rax             # Restore the syscall return stack pointer
-    mov     rax,                    rdi             # Correct syscall return value register
-
-    # Load trusted preserved registers (except rsp)
-    pop     r15
-    pop     r14
-    pop     r13
-    pop     r12
-    pop     rbp
-    pop     rbx
-
-    # Clear all argument registers
-    xor     rcx,                    rcx
-    xor     rdx,                    rdx
-    xor     rsi,                    rsi
-    xor     rdi,                    rdi
-    xor     r8,                     r8
-    xor     r9,                     r9
-
-    # Clear all temporary registers
-    xor     r10,                    r10
-    xor     r11,                    r11
-
-    # Clear CPU flags
-    add     r11,                    r11
-    cld
-
-    ret                                             # Jump to address on the stack
-    ",
-        SRSP = const SRSP,
-        options(noreturn)
-        )
-    }
-
-    #[naked]
-    pub unsafe extern "sysv64" fn syscall(aex: &mut StateSaveArea, ctx: &Context) -> u64 {
-        asm!("
-    # int syscall(rdi = aex, rsi = ctx);
-
-    # Save preserved registers (except rsp)
-    push    rbx
-    push    rbp
-    push    r12
-    push    r13
-    push    r14
-    push    r15
-
-    mov     QWORD PTR [rdi + {SRSP}], rsp             # Save restoration stack pointer
-
-    # Clear the extended CPU state
-    push    rax                                     # Save rax
-    push    rdx                                     # Save rdx
-    mov     rdx,                    ~0              # Set mask for xrstor in rdx
-    mov     rax,                    ~0              # Set mask for xrstor in rax
-
-    xrstor  [rip + {XSAVE}]                         # Clear xCPU state with synthetic state
-    pop     rdx                                     # Restore rdx
-    pop     rax                                     # Restore rax
-
-    xor     rcx,                    rcx             # Clear rcx
-
-    # Clear CPU flags
-    add     rcx,                    rcx
-    cld
-
-    mov     rsp,                    rsi             # Get exit context
-
-    jmp     {EEXIT}
-    ",
-        SRSP = const SRSP,
-        XSAVE = sym XSAVE,
-        EEXIT = sym enclu_eexit,
-        options(noreturn)
-        )
+    pub unsafe extern "sysv64" fn _start() -> ! {
```

Comment:
Actually, I think I want to go the other way around and give all the functions `extern "sysv64"` since that's the actual calling convention we implement.

Reply:
Hmm, ok, from a register standpoint on which registers to preserve.. yes

Reply:
I made all the functions `extern "sysv64"`.